algo.py

```diff
@@ -96,24 +96,17 @@ def generate_features(self, df, type="train"):
             "letters_per_word": lambda x: len(x) / (x.count(" ") + 1),
             "commas": lambda x: x.count(","),
         }
-
-        # Create a list of pandas columns.
-        # This guarantees order -- dictionaries may not.
-        hand_features_list = []
-        hand_feature_names = []
+        hand_chosen_features = pd.DataFrame()
 
         for col in ["text", "summary"]:
             for name, func in transform_functions.iteritems():
-                hand_features_list.append(df[col].apply(func))
-                hand_feature_names.append("{0}_{1}".format(col, name))
-
-        hand_chosen_features = pd.concat(hand_features_list, axis=1)
-        hand_chosen_features.columns = hand_feature_names
+                hand_chosen_features["{0}_{1}".format(col, name)] = df[col].apply(func)
```

Comment:
My question here is -- dicts don't have guaranteed ordering, so does the Dataframe call do a sort on name before creating the df?  Otherwise, each column could have a different index in the train and test pass of this method.

If you couple that with the hstack call, you could conceivably get 2 arrays with differently ordered columns for the train and test/valid sets.  Or maybe I'm missing something.